This excercise is for retail optimisation of shelf space display of products depending upon the sales lift per product on the shelf space. There are totally 8 Product types per shelf space from three different Brands .There are totally 10 shelfs.  
The constraints are that : Only one product can be selected among the 8 products.  
Also , each product type can be selected for display on the 10 shelfs at the most 'n' number of times , which can be 0 , to 1 or 2 etc.  



In [1]:
import pandas as pd
import numpy as np
import pulp

In [2]:
## Matrix of sales lift.
sales_lift = np.array([[950,818,650,848,630,648,842,842],
                [691, 849,615 ,700,653,598,563,563 ],
                [427,413,349,347,407,237,465,465],
                [345,153,282, 301,477,432, 313,313],
                [464,470,126,398,534,312,326,326],
                [281,144,283,200,168,107, 148,148],
                [238,500,291,434,465,488,544,544],
                [138,86,119,149,92,136,119,119],
                [127,124,141,54,130,140,75,75],
                [70,141,78,106,69,51,58,58]])

In [3]:
sales_lift

array([[950, 818, 650, 848, 630, 648, 842, 842],
       [691, 849, 615, 700, 653, 598, 563, 563],
       [427, 413, 349, 347, 407, 237, 465, 465],
       [345, 153, 282, 301, 477, 432, 313, 313],
       [464, 470, 126, 398, 534, 312, 326, 326],
       [281, 144, 283, 200, 168, 107, 148, 148],
       [238, 500, 291, 434, 465, 488, 544, 544],
       [138,  86, 119, 149,  92, 136, 119, 119],
       [127, 124, 141,  54, 130, 140,  75,  75],
       [ 70, 141,  78, 106,  69,  51,  58,  58]])

In [4]:
shelf_number = [1,2,3,4,5,6,7,8,9,10]

In [5]:
products = ['Unilever_1','Unilever_2','Unilever_3','Godrej_4','Godrej_5','Godrej_6','Dabur_7','Dabur_8']

In [6]:
model = pulp.LpProblem('shelf_space_opt', pulp.LpMaximize)

In [7]:
X = pulp.LpVariable.dicts('X%s_%s', (shelf_number,products),  cat = 'Binary')

In [8]:
X[1]['Unilever_1']

X1_Unilever_1

In [9]:
S = dict(zip(shelf_number,[dict(zip(products,sales_lift[0])),
                       dict(zip(products,sales_lift[1])),
                       dict(zip(products,sales_lift[2])),
                        dict(zip(products, sales_lift[3])),
                        dict(zip(products, sales_lift[4])),
                        dict(zip(products, sales_lift[5])),
                        dict(zip(products, sales_lift[6])),
                        dict(zip(products, sales_lift[7])),
                        dict(zip(products, sales_lift[8])),
                        dict(zip(products, sales_lift[9]))]))

In [10]:
S[1]

{'Dabur_7': 842,
 'Dabur_8': 842,
 'Godrej_4': 848,
 'Godrej_5': 630,
 'Godrej_6': 648,
 'Unilever_1': 950,
 'Unilever_2': 818,
 'Unilever_3': 650}

In [11]:
## Objective - to maximise sales 

sales = sum(sum(S[i][j] * X[i][j] for j in products) for i in shelf_number)
model += sales

In [12]:
## Constraints 

## Only one product per brand per shelf
for i in shelf_number:
    model += sum(X[i][j] for j in products) <= 1
    
## Number of times a product can be marketed 
no_times_markt = dict(zip(products,[1,0,0,2,2,3,1,1]))
for j in products:
    model += sum(X[i][j] for i in shelf_number) == no_times_markt[j]



In [13]:
no_times_markt

{'Dabur_7': 1,
 'Dabur_8': 1,
 'Godrej_4': 2,
 'Godrej_5': 2,
 'Godrej_6': 3,
 'Unilever_1': 1,
 'Unilever_2': 0,
 'Unilever_3': 0}

In [14]:
## save model to LP file 
model.writeLP("shelf_space_opt.Lp")

[X10_Dabur_7,
 X10_Dabur_8,
 X10_Godrej_4,
 X10_Godrej_5,
 X10_Godrej_6,
 X10_Unilever_1,
 X10_Unilever_2,
 X10_Unilever_3,
 X1_Dabur_7,
 X1_Dabur_8,
 X1_Godrej_4,
 X1_Godrej_5,
 X1_Godrej_6,
 X1_Unilever_1,
 X1_Unilever_2,
 X1_Unilever_3,
 X2_Dabur_7,
 X2_Dabur_8,
 X2_Godrej_4,
 X2_Godrej_5,
 X2_Godrej_6,
 X2_Unilever_1,
 X2_Unilever_2,
 X2_Unilever_3,
 X3_Dabur_7,
 X3_Dabur_8,
 X3_Godrej_4,
 X3_Godrej_5,
 X3_Godrej_6,
 X3_Unilever_1,
 X3_Unilever_2,
 X3_Unilever_3,
 X4_Dabur_7,
 X4_Dabur_8,
 X4_Godrej_4,
 X4_Godrej_5,
 X4_Godrej_6,
 X4_Unilever_1,
 X4_Unilever_2,
 X4_Unilever_3,
 X5_Dabur_7,
 X5_Dabur_8,
 X5_Godrej_4,
 X5_Godrej_5,
 X5_Godrej_6,
 X5_Unilever_1,
 X5_Unilever_2,
 X5_Unilever_3,
 X6_Dabur_7,
 X6_Dabur_8,
 X6_Godrej_4,
 X6_Godrej_5,
 X6_Godrej_6,
 X6_Unilever_1,
 X6_Unilever_2,
 X6_Unilever_3,
 X7_Dabur_7,
 X7_Dabur_8,
 X7_Godrej_4,
 X7_Godrej_5,
 X7_Godrej_6,
 X7_Unilever_1,
 X7_Unilever_2,
 X7_Unilever_3,
 X8_Dabur_7,
 X8_Dabur_8,
 X8_Godrej_4,
 X8_Godrej_5,
 X8_Godrej

In [15]:
## View the model 
print(model)

shelf_space_opt:
MAXIMIZE
58*X10_Dabur_7 + 58*X10_Dabur_8 + 106*X10_Godrej_4 + 69*X10_Godrej_5 + 51*X10_Godrej_6 + 70*X10_Unilever_1 + 141*X10_Unilever_2 + 78*X10_Unilever_3 + 842*X1_Dabur_7 + 842*X1_Dabur_8 + 848*X1_Godrej_4 + 630*X1_Godrej_5 + 648*X1_Godrej_6 + 950*X1_Unilever_1 + 818*X1_Unilever_2 + 650*X1_Unilever_3 + 563*X2_Dabur_7 + 563*X2_Dabur_8 + 700*X2_Godrej_4 + 653*X2_Godrej_5 + 598*X2_Godrej_6 + 691*X2_Unilever_1 + 849*X2_Unilever_2 + 615*X2_Unilever_3 + 465*X3_Dabur_7 + 465*X3_Dabur_8 + 347*X3_Godrej_4 + 407*X3_Godrej_5 + 237*X3_Godrej_6 + 427*X3_Unilever_1 + 413*X3_Unilever_2 + 349*X3_Unilever_3 + 313*X4_Dabur_7 + 313*X4_Dabur_8 + 301*X4_Godrej_4 + 477*X4_Godrej_5 + 432*X4_Godrej_6 + 345*X4_Unilever_1 + 153*X4_Unilever_2 + 282*X4_Unilever_3 + 326*X5_Dabur_7 + 326*X5_Dabur_8 + 398*X5_Godrej_4 + 534*X5_Godrej_5 + 312*X5_Godrej_6 + 464*X5_Unilever_1 + 470*X5_Unilever_2 + 126*X5_Unilever_3 + 148*X6_Dabur_7 + 148*X6_Dabur_8 + 200*X6_Godrej_4 + 168*X6_Godrej_5 + 107*X6_Godrej_

In [16]:
## Solve the model 
model.solve()

1

In [17]:
## print status and the variables 
print("Status ",pulp.LpStatus[model.status])

Status  Optimal


In [18]:
print("Objective :",pulp.value(model.objective))

Objective : 4197.0


In [19]:
results = []
for s in shelf_number:
    for p in products:
        var_output = {
        "shelf_no" : s,
        "product": p,
        "Allocation" : X[s][p].varValue
        }

        results.append(var_output)
##
shelf_space_allocation_df = pd.DataFrame.from_records(results).sort_values(["shelf_no",'product'])

print("\nShelf Space Allocation")
print("------------------------")
print(shelf_space_allocation_df.pivot(index= 'shelf_no', columns = 'product')['Allocation'])


Shelf Space Allocation
------------------------
product   Dabur_7  Dabur_8  Godrej_4  Godrej_5  Godrej_6  Unilever_1  \
shelf_no                                                               
1             0.0      0.0       0.0       0.0       0.0         1.0   
2             0.0      0.0       1.0       0.0       0.0         0.0   
3             1.0      0.0       0.0       0.0       0.0         0.0   
4             0.0      0.0       0.0       1.0       0.0         0.0   
5             0.0      0.0       0.0       1.0       0.0         0.0   
6             0.0      0.0       1.0       0.0       0.0         0.0   
7             0.0      1.0       0.0       0.0       0.0         0.0   
8             0.0      0.0       0.0       0.0       1.0         0.0   
9             0.0      0.0       0.0       0.0       1.0         0.0   
10            0.0      0.0       0.0       0.0       1.0         0.0   

product   Unilever_2  Unilever_3  
shelf_no                          
1               

In [141]:
for v in model.variables():
    if v.varValue > 0:
        print()
        print(v.name , "=", v.varValue)


X10_Godrej_6 = 1.0

X1_Unilever_1 = 1.0

X2_Godrej_4 = 1.0

X3_Dabur_7 = 1.0

X4_Godrej_5 = 1.0

X5_Godrej_5 = 1.0

X6_Godrej_4 = 1.0

X7_Dabur_8 = 1.0

X8_Godrej_6 = 1.0

X9_Godrej_6 = 1.0


Hence we can achieve a maximum sales 0f 4197 with the given constraints.

We can further work on looking at improving the profitability by knowing the cost and price per product given the sales lift.This would then change the shelf space assortment.  